feat: add smart contract example (#101)
kurtosis-tech/awesome-kurtosis: smart-contract-example/README.md

> ## Smart Contract Workflow Example

Reviewer: ```suggestion
## Ethereum dApp and Smart Contract Development Example
```

Author: sounds good!

> This example demonstrates how Kurtosis can be used for smart contract dapp development 

Reviewer: ```suggestion
This example demonstrates how Kurtosis can be used for smart contract dApp development 
```

> This example demonstrates how Kurtosis can be used for smart contract dapp development 

Reviewer: ```suggestion
This example demonstrates how Kurtosis can be used for local dApp and Smart Contract development.
```

Author: sounds good!

by using the [`eth-network-package`](https://github.com/kurtosis-tech/eth-network-package) as a local Ethereum network. 
The `eth-network package` can be used as a low overhead, configurable, and composable alternative to services like

> `hardhat-network`, `ganache` and `anvil`.

Reviewer: ```suggestion
`hardhat-network`, `ganache` and `anvil`. Kurtosis provides the developer more control and flexibility over the testnet they are using - a large reason why the [Ethereum Foundation used Kurtosis to test the Merge](https://www.kurtosis.com/blog/testing-the-ethereum-merge), and why they continue to use Kurtosis for various tests for upcoming upgrades to the network.

While this example covers only the Ethereum blockchain, Kurtosis can be used to locally configure and instantiate other blockchains (e.g. [NEAR](https://docs.near.org/develop/testing/kurtosis-localnet), [Avalanche](https://medium.com/avalancheavax/introducing-kurtosis-a-complete-testing-platform-to-accelerate-development-on-avalanche-6ad7e1147791) and allows you to connect it to any containerized service you wish.
```

Author: love this addition! helps communicate better why a dapp developer might choose us over other options.

### Setup

> This folder contains a typical setup for a smart contract developer looking to develop a dapp using the 

Reviewer: ```suggestion
This folder contains a typical setup for a dApp developer using the
```

[Hardhat](https://hardhat.org/) framework. Let's explore it's contents.

> - `contracts/` contains a few simple smart contracts for a Blackjack dapp 

Reviewer: ```suggestion
- `contracts/` contains a few simple smart contracts for a Blackjack dApp 
```

Author: sounds good

- `scripts/` contains a script to deploy a token contract to our local Ethereum network
- `test/` contains a simple test for our token contract
- `hardhat.config.ts` configures our Hardhat setup. 
It allows us to configure Hardhat to use a local Ethereum network created by the `eth-network-package`.

### Running the Example

> This assumes you have the following services installed:

Reviewer: Probably should add Docker here too because you need Docker to use Kurtosis. Specifically installing Docker and starting up the Docker Daemon. I'd recommend linking to our docs about this here https://docs.kurtosis.com/install#i-install--start-docker

                                                                  metrics: 5054/tcp -> http://127.0.0.1:64257   
                                                                  tcp-discovery: 9000/tcp -> 127.0.0.1:64258    
                                                                  udp-discovery: 9000/udp -> 127.0.0.1:59023    
    a4429f19d246   cl-client-0-validator                          http: 5042/tcp -> 127.0.0.1:64262             RUNNING
                                                                  metrics: 5064/tcp -> http://127.0.0.1:64263   
    77c853ab371b   el-client-0                                    engine-rpc: 8551/tcp -> 127.0.0.1:64246       RUNNING
                                                                  rpc: 8545/tcp -> 127.0.0.1:64248              
                                                                  tcp-discovery: 30303/tcp -> 127.0.0.1:64247   
                                                                  udp-discovery: 30303/udp -> 127.0.0.1:58705   
                                                                  ws: 8546/tcp -> 127.0.0.1:64249               

>     0d8469f528c9   prelaunch-data-generator-1680298734419927594   <none>                                        STOPPED

Reviewer: What are these prelaunch data generators, why are they important/used, and why are they stopped? Would love a sentence or two about that!

Author: Great idea.

    8828574730f6   prelaunch-data-generator-1680298734428736261   <none>                                        STOPPED
    a9ad4e4cc65c   prelaunch-data-generator-1680298734439099469   <none>                                        STOPPED
    ```

>     We see a single node with a geth EL client and lighthouse CL client running has been created. The network configuration is configurable via a json file. Read [here](https://github.com/kurtosis-tech/eth-network-package#configuring-the-network) to learn more.

Reviewer: ```suggestion
    We see a single node with a geth EL client and lighthouse CL client running has been created. The CL and EL client pair can be configured using a json file. Currently, the Ethereum package supports lighthouse and nimbus CL clients as well as the erigon and geth EL clients. Read [here](https://github.com/kurtosis-tech/eth-network-package#configuring-the-network) to learn more.
```

Author: sounds good

> 2. Replace `<PORT>` in `hardhat.config.ts` with the port of the rpc uri output from any `el-client-` service. In this case, the port would be `64248`.

Reviewer: What are the different types of `el-client-` that a user would see?

Author: The `eth-network-package` labels each el client service with the name`el-client-<num>` so eg. `el-client-0`, `el-client-1`, `el-client-2`.... In this example since we only spin one node, we see `el-client-0`. Is there a way I can make this more clear? Maybe `el-client-<num>` instead of `el-client-`?

Reviewer: Cool that makes sense. `el-client-<num>` and `cl-client-<num>` makes more sense, with `<num>` equal to the instance of that client (whether its EL or CL)

    ```
    localnet: {
    url: 'http://127.0.0.1:<PORT>',//TODO: REPLACE PORT WITH THE PORT OF A NODE URI PRODUCED BY THE ETH NETWORK KURTOSIS PACKAGE

>     gasPrice: 225000000000,

Reviewer: Can we add a comment here that this is the default gasPrice in Wei?
```suggestion
    gasPrice: 225000000000, # Default gasPrice, in wei, from Hardhat
```

## Ethereum dApp and Smart Contract Development Example

> This example demonstrates how Kurtosis can be used for local dApp and Smart Contract development.

Reviewer: ```suggestion
This example demonstrates how Kurtosis can be used for local dApp and Smart Contract development
```

by using the [`eth-network-package`](https://github.com/kurtosis-tech/eth-network-package) as a local Ethereum network. 

> The `eth-network package` can be used as a low overhead, configurable, and composable alternative to services like

Reviewer: ```suggestion
The `eth-network package` can be used as a low overhead, configurable, and composable alternative to frameworks like
```

`hardhat-network`, `ganache` and `anvil`. Kurtosis provides the developer more control and flexibility over the testnet they are using - a large reason why the [Ethereum Foundation used Kurtosis to test the Merge](https://www.kurtosis.com/blog/testing-the-ethereum-merge), and why they continue to use Kurtosis for various tests for upcoming upgrades to the network.

> While this example covers only the Ethereum blockchain, Kurtosis can be used to locally configure and instantiate other blockchains (e.g. [NEAR](https://docs.near.org/develop/testing/kurtosis-localnet), [Avalanche](https://medium.com/avalancheavax/introducing-kurtosis-a-complete-testing-platform-to-accelerate-development-on-avalanche-6ad7e1147791) and allows you to connect it to any containerized service you wish.

Reviewer: ```suggestion
While this example covers only the Ethereum blockchain, Kurtosis can be used to locally configure and instantiate other blockchains (e.g. [NEAR](https://docs.near.org/develop/testing/kurtosis-localnet), [Avalanche](https://medium.com/avalancheavax/introducing-kurtosis-a-complete-testing-platform-to-accelerate-development-on-avalanche-6ad7e1147791) and allows you to connect your local testnet with any containerized service you wish.
```

      a4429f19d246   cl-client-0-validator                          http: 5042/tcp -> 127.0.0.1:64262             RUNNING
                                                                    metrics: 5064/tcp -> http://127.0.0.1:64263   
      77c853ab371b   el-client-0                                    engine-rpc: 8551/tcp -> 127.0.0.1:64246       RUNNING
                                                                    rpc: 8545/tcp -> 127.0.0.1:64248              
                                                                    tcp-discovery: 30303/tcp -> 127.0.0.1:64247   
                                                                    udp-discovery: 30303/udp -> 127.0.0.1:58705   
                                                                    ws: 8546/tcp -> 127.0.0.1:64249               
      0d8469f528c9   prelaunch-data-generator-1680298734419927594   <none>                                        STOPPED
      8828574730f6   prelaunch-data-generator-1680298734428736261   <none>                                        STOPPED
      a9ad4e4cc65c   prelaunch-data-generator-1680298734439099469   <none>                                        STOPPED
      ```

>       We see a single node with a geth EL client and lighthouse CL client running has been created. The CL and EL client pair can be configured using a json file. Currently, the Ethereum package supports lighthouse, nimbus, lodestar, teku, and prysm CL clients as well as the erigon, nethermind, besu, and geth EL clients. Read [here](https://github.com/kurtosis-tech/eth-network-package#configuring-the-network) to learn more. 

Reviewer: ```suggestion
      We see a single node with a geth EL client and lighthouse CL client running has been created. The CL and EL client pair can be configured using a `.json` file. Currently, the Ethereum package supports lighthouse, nimbus, lodestar, teku, and prysm CL clients as well as the erigon, nethermind, besu, and geth EL clients. Read [here](https://github.com/kurtosis-tech/eth-network-package#configuring-the-network) to learn more. 
```

>       Each EL and CL client require data that differs per client, so we leverage the `prelaunch-data-generator` to create it, built off this [Docker image](https://github.com/ethpandaops/ethereum-genesis-generator).

Reviewer: > so we leverage the `prelaunch-data-generator` to create it

When you say "it", are you referring to the data? What kind of data are we generating?

        "48fcc39ae27a0e8bf0274021ae6ebd8fe4a0e12623d61464c498900b28feb567",
        "7988b3a148716ff800414935b305436493e1f25237a2a03e5eebc343735e2f31",
        "b3c409b6b0b3aa5e65ab2dc1930534608239a478106acf6f3d9178e9f9b00b35",
        "df9bb6de5d3dc59595bcaa676397d837ff49441d211878c024eabda2cd067c9f",
        "7da08f856b5956d40a72968f93396f6acff17193f013e8053f6fbb6c08c194d6",
      ],
    },
    ```
3. Run 
   ```
   npx hardhat balances --network localnet
   ``` 
   This verifies that network is working and detects the prefunded accounts on the network, created by the `eth-network-package`.
   The output should look something like this
    ```
    0x878705ba3f8Bc32FCf7F4CAa1A35E72AF65CF766 has balance 10000000000000000000000000
    0x4E9A3d9D1cd2A2b2371b8b3F489aE72259886f1A has balance 10000000000000000000000000
    0xdF8466f277964Bb7a0FFD819403302C34DCD530A has balance 10000000000000000000000000
    0x5c613e39Fc0Ad91AfDA24587e6f52192d75FBA50 has balance 10000000000000000000000000
    0x375ae6107f8cC4cF34842B71C6F746a362Ad8EAc has balance 10000000000000000000000000
    0x1F6298457C5d76270325B724Da5d1953923a6B88 has balance 10000000000000000000000000
    ```

> 4. Now, we can run dev/test workflows against our network! For example, let's compile and deploy the `ChipToken` 

Reviewer: ```suggestion
4. Now, we can run dev/test workflows against our network! For example, let's compile and deploy the `ChipToken` so we can iterate and test how things work locally:
```
